### ISSUE TITLE
png thumbnail memory issue

### ISSUE THREAD

**Issue by juyunsang** [jcupitt/libvips#646]
Good morning
Memory usage continues to increase during the creation of certain png image thumbnail.
There is an issue that the system kills while the memory usage is increasing.
I debugged the memory-intensive libvips api.
Vips_image_write_to_file (rotateOut, input-> outputFilePath, NULL) increases memory usage.
I use libpng version 0.16 and libvips version 8.5.
I will attach it to the problem image file.
Thank you.

![test](https://cloud.githubusercontent.com/assets/13113520/25424811/c058a2f0-2aa4-11e7-9bbb-eaa9ff87a2bb.png)

**Comment by jcupitt** [comment]
Hello, everything will struggle with large interlaced PNG images. These images cannot be streamed: you must read the entire image into memory before you can start processing.

https://en.wikipedia.org/wiki/Adam7_algorithm

For a 10,000 x 10,000 interlaced PNG, you must allocate at least 400mb of memory :-( I don't think there's anything that can be done about this, unfortunately, except to beg people not to use this ridiculous format.

**Comment by jcupitt** [comment]
Ooops hehe now I look at the image in detail, I think it's a bomb. If you check the header:

```
$ vipsheader x.png 
x.png: 787402x787402 uchar, 4 bands, srgb, pngload
```

But of course the image is only 1.2mb. It looks like it has been truncated. I wonder why read doesn't stop on the first error? I'll have a look.

Before processing, I would open the image and check the size. If either axis is over (for example) 100,000 pixels, I would reject the image.

**Comment by jcupitt** [comment]
I found a problem with png reading for very large, truncated files, thank you for the report. 

I now see:

```
$ vipsthumbnail x.png[fail]
(vipsthumbnail:30628): VIPS-WARNING **: error in tile 0 x 400
vipsthumbnail: unable to thumbnail x.png[fail]
vipspng: Read Error
```

The `[fail]` after the filename means stop on the first error.

This fix will be in 8.5.5.

**Comment by juyunsang** [comment]
Good morning
Thank you for quick response.
8.5.5 What is the release schedule?
Can you apply to the master branch in a hurry?
Thank you.

**Comment by jcupitt** [comment]
Yes, it's in master now. I'd guess 8.5.5 would be another few weeks.

**Comment by juyunsang** [comment]
Good morning
I have installed 8.5 branch and found that pkg-config --modversion vips is version 8.5.5.
I have created a problem png file thumbnail, but the following error message continues to occur.

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

(Vipsthumbnail: 119695): VIPS-WARNING **: Read Error

....
Have not bugs been fixed in 8.5 branch yet?

**Comment by jcupitt** [comment]
Hello again, what did you type to get that message? I see:

```
$ head -c 10000 k2.png > x.png
$ vips copy x.png[fail] y.png

(vips:21010): VIPS-WARNING **: Read Error

(vips:21010): VIPS-WARNING **: error in tile 0 x 0
vipspng: libpng read error
vips2png: unable to write "y.png"
$
```

You need the `fail` option or libvips will try to read the whole image. If you are using C, you need:

```
VipsImage *image = vips_image_new_from_file( "somefile.png", 
    "fail", TRUE,
    NULL ); 
```

**Comment by juyunsang** [comment]
Good morning.
Thank you for answer.
Previously I read vips image vips_image_new_from_file (input-> inputFilePath, "access", VIPS_ACCESS_RANDOM, NULL); And vips_image_new_from_file (input-> inputFilePath, "access", VIPS_ACCESS_SEQUENTIAL_UNBUFFERED, NULL).
Vips_image_new_from_file ("somefile.png", "fail", TRUE, NULL); I made this change and checked the error in the png file.
But throwing error image to the Thumbnail generation following that success in existing and Thumbnail creation failed.

(Vips: 54946): VIPS-WARNING **: error in tile 0 x 0

(Vips: 54946): VIPS-WARNING **: error in tile 0 x 8
VipsJpeg: Invalid SOS parameters for sequential JPEG
VipsJpeg: out of order read at line 0

**Comment by jcupitt** [comment]
Sorry @juyunsang, I don't understand.

What do you want to happen when you try to process a corrupted PNG file?

**Comment by juyunsang** [comment]
Good morning
I am sorry that I can not speak English well.
Corrupted png processing is perfect.
However, before the attached image has the [fail] option, the thumbnail creation succeeds,
Thumbnail creation will fail if [fail] option is entered.
Thank you for answer.
Have a nice day
![3](https://cloud.githubusercontent.com/assets/13113520/25891222/32a9d124-35ab-11e7-93e5-8d8ea951d165.jpg)

**Comment by jcupitt** [comment]
But what is the problem? Can you describe the behaviour you want?

**Comment by juyunsang** [comment]
Good morning.
I am sorry that I am not fluent in English.
Corrupted png processing is perfect.
However, the attached image fails to create the thumbnail when the fail option is added.
If the fail option is removed, the thumbnail is created normally.
It seems to be a bug and I would appreciate your confirmation.

issue jpeg image
![issue_image](https://cloud.githubusercontent.com/assets/13113520/25977629/b157c7e0-36f7-11e7-91ea-2ad5efac7794.jpg)

**Comment by juyunsang** [comment]
Good morning.
I am sorry that I am not fluent in English.
Corrupted png processing is perfect.
However, the attached image fails to create the thumbnail when the fail option is added.
If the fail option is removed, the thumbnail is created normally.
It seems to be a bug and I would appreciate your confirmation.

issue jpeg image
![issue_image](https://cloud.githubusercontent.com/assets/13113520/25977629/b157c7e0-36f7-11e7-91ea-2ad5efac7794.jpg)

**Comment by jcupitt** [comment]
Ah I was confused, I thought we were still talking about PNG images. I'll have a look.

**Comment by jcupitt** [comment]
Yes, you are right, the jpeg loader was treating warnings as errors. I've pushed a fix to the 8.5 branch and it seems to be better now. Thanks!
